manticore/platforms/evm_world_state.py

```diff
@@ -0,0 +1,431 @@
+        pass
+
+    @abstractmethod
+    def get_timestamp(self) -> Union[int, BitVec]:
+        pass
+
+    @abstractmethod
+    def get_difficulty(self) -> Union[int, BitVec]:
+        pass
+
+    @abstractmethod
+    def get_gaslimit(self) -> Union[int, BitVec]:
+        pass
+
+    @abstractmethod
+    def get_coinbase(self) -> Union[int, BitVec]:
+        pass
+
+
+####################################################################################################
+
+
+class DefaultWorldState(WorldState):
+    def accounts(self) -> Set[int]:
+        return set()
+
+    def get_nonce(self, address: int) -> int:
+        return 0
+
+    def get_balance(self, address: int) -> int:
+        return 0
+
+    def has_storage(self, address: int) -> bool:
+        return False
+
+    def get_storage(self, address: int) -> Optional[Storage]:
+        raise NotImplementedError
+
+    def get_storage_data(self, address: int, offset: Union[int, BitVec]) -> int:
+        return 0
+
+    def get_code(self, address: int) -> bytes:
+        return bytes()
+
+    def get_blocknumber(self) -> int:
+        # assume initial byzantium block
+        return 4370000
+
+    def get_timestamp(self) -> int:
+        # 1524785992; // Thu Apr 26 23:39:52 UTC 2018
+        return 1524785992
+
+    def get_difficulty(self) -> int:
+        return 0
+
+    def get_gaslimit(self) -> int:
+        return 0
+
+    def get_coinbase(self) -> int:
+        return 0
+
+
+####################################################################################################
+
+
+class Endpoint:
+    def __init__(self, blocknumber: int, warned: bool):
+        self.blocknumber = blocknumber
+        self.warned = warned
+
+
+_endpoints: Dict[str, Endpoint] = {}
+
+
+def _web3_address(address: int) -> ChecksumAddress:
+    return Web3.toChecksumAddress("0x%0.40x" % address)
+
+
+# sam.moelius: Notes:
+#
+# 1. https://github.com/ethereum/wiki/wiki/JSON-RPC lists the kinds of information that an Ethereum
+#    node can provide over JSON RPC.
+#
+# 2. The accounts and get_storage methods do not make sense when using JSON RPC.  IMHO, we should
+#    program to the least common denominator.  In that sense, we should see whether the accounts and
+#    get_storage methods could be eliminated.
+
+
+class RemoteWorldState(WorldState):
+    def __init__(self, url: str):
+        actual = urlparse(url)
+        expected = ParseResult(scheme="", netloc="", path=url, params="", query="", fragment="")
+        if actual != expected:
+            raise EthereumError("URL must be of the form 'IP:PORT': " + url)
+        self._url = url
+
+    def _web3(self) -> Web3:
+        # sam.moelius: Force WebsocketProvider.__init__ to call _get_threaded_loop.  The existing
+        # "threaded loop" could be leftover from a fork, in which case it will not work.
+        Web3.WebsocketProvider._loop = None
+        web3 = Web3(Web3.WebsocketProvider(URI("ws://" + self._url)))
```

Comment:
Does this handle the case of a valid URL that doesn't point to an ethereum node?

Reply:
Not sure. What do you think Manticore should do in that case?

Reply:
As it is, I'd expect that feeding this a bad URL would throw an exception and make Manticore crash. Rather than crashing, I think Manticore should catch that exception, do any necessary cleanup, and then raise an `EthereumError`, similar to what it does for an improperly formatted URL. 


> What if a valid URL pointed to something that spoke the protocol, but returned bogus results?

Hmm, I don't know of a way to handle this, but it doesn't sound to me like we need to. On the WASM side, though there's a lot of static validation Manticore could do on the binaries, I take the stance of "If you give me a bad binary, that's _your_ problem." Somewhat similarly, I think handling the easy failure cases relatively gracefully is enough here.

Reply:
A bad URL now causes Manticore to raise an `EthereumError`.